File: sem4/jpp/l3/main.cpp
#include "DHSetup.hpp"
#include "User.hpp"

#define CHARACTERISTIC 1234567891

int main() {
    DHSetup<CHARACTERISTIC> dhSetup;
    GF_Int<CHARACTERISTIC> generator = dhSetup.getGenerator();
    GF_Int<CHARACTERISTIC> powerResult = dhSetup.power(generator, 5);

    // std::cout << "Generator: " << generator.get_value() << std::endl;
    // for(int i = 0; i < generator.get_characteristic() - 1; i++) {
    //     std::cout << "Generator^" << i << ": " << dhSetup.power(generator, i).get_value() << "\n";
    // }

    User<CHARACTERISTIC> alice(dhSetup);
    User<CHARACTERISTIC> bob(dhSetup);

    alice.setKey(bob.getPublicKey());
    bob.setKey(alice.getPublicKey());

    GF_Int<CHARACTERISTIC> message(6969420);
    auto encryptedMessage = alice.encrypt(message);
    auto decryptedMessage = bob.decrypt(encryptedMessage);

    std::cout << "Wiadomosc oryginalna: " << message.get_value() << "\n";
    std::cout << "Encrypted Message: " << encryptedMessage.get_value() << "\n";
    std::cout << "Decrypted Message: " << decryptedMessage.get_value() << "\n";

    return 0;
}

File: sem4/jpp/l3/java/Main.java
public class Main {
    public static void main(String[] args) {
        DHSetup dhSetup = new DHSetup(1234567891);
        System.out.println("Generator: " + dhSetup.getGenerator());
        System.out.print("Is generator actually a generator?: ");
        System.out.println(dhSetup.isGenerator(dhSetup.getGenerator().getValue(), 1234567891));
        User alice = new User(dhSetup);
        User bob = new User(dhSetup);
        System.out.println("Alice secret: " + alice.getSecret());
        System.out.println("Bob secret: " + bob.getSecret());
        GF_Int alicePublicKey = bob.getPublicKey();
        GF_Int bobPublicKey = alice.getPublicKey();
        System.out.println("Alice's public key: " + alicePublicKey);
        System.out.println("Bob's public key: " + bobPublicKey);

        alice.setKey(bobPublicKey);
        bob.setKey(alicePublicKey);

        GF_Int message = new GF_Int(4206969, dhSetup.getCharacteristic()); 
        GF_Int encryptedMessage = alice.encrypt(message); 
        GF_Int decryptedMessage = bob.decrypt(encryptedMessage); 

        System.out.println("Original Message: " + message);
        System.out.println("Encrypted Message: " + encryptedMessage);
        System.out.println("Decrypted Message: " + decryptedMessage);
    }
}